api.ts

```diff
@@ -1423,23 +1440,46 @@ export function makeMetadataFormula(
   });
 }
 
-function makePropertyAutocompleteFormula(
-  execute: PropertyAutocompleteMetadataFunction,
-  options?: {connectionRequirement?: ConnectionRequirement},
-): PropertyAutocompleteMetadataFormula {
+/**
+ * Builds a formula to store in {@link SyncTableAutocompleters}.
+ *
+ * @hidden
+ */
+export function makePropertyAutocompleteFormula<SchemaT extends Schema>({
+  execute,
+  schema,
+  name,
+}: {
+  execute: PropertyAutocompleteMetadataFunction<Array<SchemaType<SchemaT>>>;
+  schema: SchemaT;
+  name: string;
+}): PropertyAutocompleteMetadataFormula<SchemaT> {
   if (!(execute instanceof Function)) {
-    throw new Error(`Value for propertyAutocomplete must be a function`);
+    throw new Error(`Value for execute must be a function`);
   }
-  return makeObjectFormula({
-    name: 'getPropertyAutocompleteMetadata',
-    description: 'Gets property autocomplete',
-    execute([], context) {
-      return execute(context as PropertyAutocompleteExecutionContext);
-    },
+
+  type ResultT = SchemaType<ArraySchema<SchemaT>>;
+
+  // The type SchemaType<ArraySchema<T>> is equivalent to Array<SchemaType<T>>, but typescript doesn't know
+  // that unless we do a cast.
+  const executeRetyped = execute as PropertyAutocompleteMetadataFunction<SchemaType<ArraySchema<SchemaT>>>;
+
+  const innerExecute = async ([]: ParamValues<[]>, context: ExecutionContext): Promise<ResultT> => {
+    const result = await executeRetyped(context as PropertyAutocompleteExecutionContext);
+    return result;
+  };
+
+  const formulaDefn: ArrayFormulaDef<[], SchemaT> = {
+    connectionRequirement: ConnectionRequirement.Optional,
```

Comment:
Any reason this shouldn't just inherit the connection requirement from e.g. the builder, via a parameter to this helper?

Reply:
In terms of why the code looks like this, the answer is copy-paste.

I'll update the PR to keep the connection requirements for autocomplete formulas updated in the same way as other metadata formulas on a sync table

Reply:
The code is now updated to set the connection requirement based on the parent sync table, with two unit tests to check behavior depending on whether `setUserAuthentication()` is called before or after `addSyncTable()`